colossalai/kernel/cuda_native/fmha/fmha.py

```diff
@@ -0,0 +1,97 @@
+import math
+from typing import Optional
+
+import torch
+import torch.nn.functional as F
+from einops import rearrange
+
+from ..scaled_softmax import AttnMaskType
+from .flash_attn_2 import HAS_FLASH_ATTN
+from .mem_eff_attn import HAS_MEM_EFF_ATTN
+from .utils import Repad, SeqLensInfo, Unpad
+
+
+class ColoAttention(torch.nn.Module):
+
+    def __init__(self, embed_dim: int, num_heads: int, dropout: float = 0.0, scale=None):
+        super().__init__()
+        assert embed_dim % num_heads == 0, \
+            f"the embed dim ({embed_dim}) is not divisible by the number of attention heads ({num_heads})."
+        if scale is not None:
+            self.scale = scale
+        else:
+            self.scale = 1 / math.sqrt(embed_dim // num_heads)
+        self.dropout = dropout
+
+        if not HAS_MEM_EFF_ATTN and not HAS_FLASH_ATTN:
+            raise Exception("flash attention can not support!")
+
+    def get_seq_info_from_mask(self, attn_mask: torch.Tensor, device="cuda"):
```

Comment:
do this calculation in SeqLenInfo.materialize, then we don't need this method in ColoAttention any more

Reply:
> do this calculation in SeqLenInfo.materialize, then we don't need this method in ColoAttention any more

updated.